extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/cdi/WithSpanInterceptor.java

```diff
@@ -53,35 +65,108 @@ public Object span(final ArcInvocationContext invocationContext) throws Exceptio
                 invocationContext.getParameters(),
                 invocationContext.getInterceptorBindings());
 
+        final Class<?> returnType = invocationContext.getMethod().getReturnType();
         Context parentContext = Context.current();
-        Context spanContext = null;
-        Scope scope = null;
         boolean shouldStart = instrumenter.shouldStart(parentContext, methodRequest);
-        if (shouldStart) {
-            spanContext = instrumenter.start(parentContext, methodRequest);
-            scope = spanContext.makeCurrent();
-        }
-
-        try {
-            Object result = invocationContext.proceed();
 
-            if (shouldStart) {
-                instrumenter.end(spanContext, methodRequest, null, null);
-            }
+        if (!shouldStart) {
+            return invocationContext.proceed();
+        }
 
-            return result;
-        } catch (Throwable t) {
-            if (shouldStart) {
-                instrumenter.end(spanContext, methodRequest, null, t);
-            }
-            throw t;
-        } finally {
-            if (scope != null) {
-                scope.close();
+        if (isUni(returnType)) {
+            final Context currentSpanContext = instrumenter.start(parentContext, methodRequest);
+            final Scope currentScope = currentSpanContext.makeCurrent();
+            return ((Uni<Object>) invocationContext.proceed()).onTermination()
+                    .invoke(new Functions.TriConsumer<Object, Throwable, Boolean>() {
+                        @Override
+                        public void accept(Object o, Throwable throwable, Boolean aCancelled) {
+                            try {
+                                if (aCancelled) {
+                                    instrumenter.end(currentSpanContext, methodRequest, null,
+                                            new CancellationException("Cancelled"));
```

Comment:
Do you need to pass a String? A cancellation exception is only thrown on cancellation

Reply:
yes, no need